### ISSUE TITLE
Generic strings/regex patterns in strings

### ISSUE THREAD

**Issue by Fidget-Spinner** [python/typing#1202]
Currently, we have `str` and `LiteralStr`. These don't do a good job expressing the range of values a string may have. Let's say we have the following function:

```python
Email: TypeAlias = str

def send_email(email: Email) -> int:
    # Send an email
    ...

# passes
send_email("test@example.com")

# passes
send_email("badddd")
```

We could improve it by having something like this:
```python
# This is a bad regex pattern, but you get the idea :)
Email: TypeAlias = str['[a-z0-9]+@[a-z0-9]\.[a-z0-9]']

def send_email(email: Email) -> int:
    # Send an email
    ...

# passes
send_email("test@example.com")

# fails
send_email("badddd")
```

A static type checker would be able to validate strings passed in by code. I would imagine this idea can be extended to `Pattern` and `Match` generics as well, but I haven't thought too deeply about them yet.

**Comment by henribru** [comment]
There's kind of some prior art in Typescript: https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html

It's not quite as powerful as this regex idea though.

**Comment by gvanrossum** [comment]
The TypeScript feature allows more than type checking though -- IIRC it allows constructing new (string literal) types from other strings.

I personally don't think a feature to let type checkers use regex matching on literals is all that useful -- I'd rather use `Email = NewType(str)` and leave the validation to runtime code that the type checker doesn't have to be understand. (It's likely that you already have an email validation routine in your system, and it may not be easy to replicate its exact functionality as a regex.)

**Comment by juanmirocks** [comment]
For reference, Pydantic (V2) supports & validates such regex-constrained, annotated strings: https://docs.pydantic.dev/2.4/api/types/#pydantic.types.StringConstraints

**Comment by mikeshardmind** [comment]
To add another detraction to this, regex in annotations to be checked *statically* is probably not a great idea for similar reasons that type checkers don't evaluate expressions. Pathological patterns, maliciously added or not, could lead to ReDoS in many regex engines, including the one in python's standard library. This could lead to merely checking out a PR causing analysis to take undue time if not specifically using a more constrained regex engine.

**Comment by Fidget-Spinner** [comment]
Yeah I've been thinking of scaling this down quite a bit and following TypeScript's types in string template style.

**Comment by dan1hc** [comment]
Resurrecting things... (apologies for posting in wrong location earlier, thanks @AlexWaygood for the `301` here : ).

### From my original post

Could `str` methods take a `TypeVar('string', bound=str)` rather than `str` itself every time? (i.e. eliminate the need for a `# type: ignore[misc]` below).

> Also, I think it might be cool to have generic `str[Any]`, such that the param could be used to indicate string casing.

### Piggybacking...

To expand a bit more on some of the concerns outlined above....

Maybe this could be done in such a way that aliases are kept for common `Casings` (i.e. `lowerCamelCase`, `snake_case`, `PascalCase`, `kebab-case` etc.)?

Then each `Casing` might have accompanying, ReDos hardened regex pre-compiled?

Personal use case would essentially be validating / converting REST requests arriving in `lowerCamelCase` to `dataclasses` style objects, for which all fields must be `snake_case`... i.e. no string should really ever be longer than `{1, 128}` in this case (and I would argue `128` is still a push for property names for any RESTful resource)... but this does seem like it could extend to `str[email]` etc. as well (same logic around the hardening... if someone is sending a 200 character long email address... that feels like something that shouldn't validate anyways?)

### Original post idea

```python
import typing as t

from . import typ


class DictLike:

    ___dataclass_fields__: dict[str[snake_case], typ.AnyField[t.Any]]

    @classmethod
    def keys(cls) -> t.KeysView[str[snake_case]]:
        """
        Return an iterator of keys like a `dict`.

        ---

        Removes any suffixed underscores from field names (`_`).

        """

        return t.KeysView(
            {
                k.rstrip('_'): v  # type: ignore[misc]
                for k, v
                in cls.__dataclass_fields__.items()
                }
            )

```

**Comment by dan1hc** [comment]
@InSyncWithFoo that's pretty sweet, definitely simpler than what I ended up doing, ha!

Ultimately, went structural: [https://github.com/dan1hc/fqr/blob/main/src/fqr/core/typ/obj.py#L206](https://github.com/dan1hc/fqr/blob/main/src/fqr/core/typ/obj.py#L206)

So I could get:

![Screenshot 2024-07-27 080735](https://github.com/user-attachments/assets/60d6e72e-410e-42e5-95b4-3fc34c50a25d)

Re Unicode characters: my use case is purely for [politely] coercing developers using my program into RESTful compliance, so the validation gets a bit easier when you add the constraint that `camelCase | snake_case` strings should never contain any `[^a-zA-Z0-9_]` to begin with [at least in the world of RESTful resources and records I would want to interact with ha].
